### PR TITLE
[FIRRTL] Add module inlining pass

### PR DESCRIPTION
**Pull request by youngar**
The module inlining pass handles module flattening and instance
inlining.  Module flattening is recursively inlining every instance in a
module.  Module inlining is inlining only the instances of  modules
which are marked for inlining.

### CODE REVIEW

File: lib/Dialect/FIRRTL/Transforms/ModuleInliner.cpp [hunk 1]
```diff
@@ -0,0 +1,394 @@
...
+using namespace circt;
+using namespace firrtl;
+
+//===----------------------------------------------------------------------===//
+// Module Inlining Support
+//===----------------------------------------------------------------------===//
+
+/// If this operation or any child operation has a name, add the prefix to that
+/// operation's name.
+static void rename(const Twine &prefix, Operation *op) {
```
Comment (lattner) [inline lib/Dialect/FIRRTL/Transforms/ModuleInliner.cpp:30]: The pervasive use of twine arguments here makes me concerned about dangling pointer bugs that happen due to twine temporaries.  Would you lose much by having these take StringRef prefixes instead?

File: lib/Dialect/FIRRTL/Transforms/ModuleInliner.cpp [hunk 2]
```diff
@@ -0,0 +1,394 @@
...
+  // TODO: this should really be an OpInterface.. Or we can just blindly
+  // modify the attribute "name" indiscriminantly.
+  llvm::TypeSwitch<Operation *>(op)
+      .Case<CMemOp, InstanceOp, MemOp, MemoryPortOp, NodeOp, RegOp, RegResetOp,
+            SMemOp, WireOp>([&](auto op) {
+        // TODO: one day names won't be optional
```
Comment (lattner) [inline lib/Dialect/FIRRTL/Transforms/ModuleInliner.cpp:36]: Is there a bug tracking this? Seems like a high priority for the current sprint to get these representational issues sorted.

File: lib/Dialect/FIRRTL/Transforms/ModuleInliner.cpp [hunk 3]
```diff
@@ -0,0 +1,394 @@
...
+        if (auto name = op.nameAttr()) {
+          std::string newName = (prefix + name.getValue()).str();
+          op.nameAttr(StringAttr::get(op.getContext(), newName));
```
Comment (lattner) [inline lib/Dialect/FIRRTL/Transforms/ModuleInliner.cpp:39]: StringAttr::get is the thing that should take a Twine! :-)

File: lib/Dialect/FIRRTL/Transforms/ModuleInliner.cpp [hunk 4]
```diff
@@ -0,0 +1,394 @@
...
+  for (auto &region : op->getRegions())
+    for (auto &block : region)
+      for (auto &op : block)
+        rename(prefix, &op);
+}
+
+/// Clone an operation, mapping used values and results with the mapper, and
+/// apply the prefix to the name of the operation. This will clone to the
+/// insert point of the builder.
+static void cloneAndRename(const Twine &prefix, OpBuilder b,
```
Comment (lattner) [inline lib/Dialect/FIRRTL/Transforms/ModuleInliner.cpp:52]: I'd recommend passing the OpBuilder by reference here.

File: lib/Dialect/FIRRTL/Transforms/ModuleInliner.cpp [hunk 5]
```diff
@@ -0,0 +1,394 @@
...
+/// between module ports and instance results. For every port in the target
+/// module, create a wire, and assign a mapping from each module port to the
+/// wire. When the body of the module is cloned, the value of the wire will be
+/// used instead of the module's ports.
+static SmallVector<Value> mapPortsToWires(const Twine &prefix, OpBuilder &b,
+                                          BlockAndValueMapping &mapper,
+                                          FModuleOp target) {
+  SmallVector<Value> wires;
+  auto portInfo = target.getPorts();
+  for (unsigned i = 0; i < target.getNumArguments(); ++i) {
```
Comment (lattner) [inline lib/Dialect/FIRRTL/Transforms/ModuleInliner.cpp:68]: nit, no need to evaluate getNumArguments() each time through the loop

File: lib/Dialect/FIRRTL/Transforms/ModuleInliner.cpp [hunk 6]
```diff
@@ -0,0 +1,394 @@
...
+
+/// This function is used after inlining a module, to handle the conversion
+/// between module ports and instance results. This maps each wire to the
+/// result of the instance operation.  When future operations are cloned from
+/// the current block, they will use the value of the wire instead of the
+/// instance results.
+static void mapResultsToWires(BlockAndValueMapping &mapper,
+                              SmallVectorImpl<Value> &wires,
+                              InstanceOp instance) {
+  for (unsigned i = 0; i < instance.getNumResults(); ++i) {
```
Comment (lattner) [inline lib/Dialect/FIRRTL/Transforms/ModuleInliner.cpp:86]: likewise

File: lib/Dialect/FIRRTL/Transforms/ModuleInliner.cpp [hunk 7]
```diff
@@ -0,0 +1,394 @@
...
+    }
+
+    // Create the wire mapping for results + ports. We RAUW the results instead
+    // of mapping them.
+    BlockAndValueMapping mapper;
+    OpBuilder b(instance);
+    auto instanceName = instance.nameAttr().getValue();
+    auto nestedPrefix = (instanceName + "_").str();
+    auto wires = mapPortsToWires(nestedPrefix, b, mapper, target);
+    for (unsigned i = 0; i < instance.getNumResults(); ++i)
```
Comment (lattner) [inline lib/Dialect/FIRRTL/Transforms/ModuleInliner.cpp:240]: same nit

File: lib/Dialect/FIRRTL/Transforms/ModuleInliner.cpp [hunk 8]
```diff
@@ -0,0 +1,394 @@
...
```
Comment (lattner) [inline lib/Dialect/FIRRTL/Transforms/ModuleInliner.cpp:324]: Here and in a few other places too

File: lib/Dialect/FIRRTL/Transforms/ModuleInliner.cpp [hunk 9]
```diff
@@ -0,0 +1,394 @@
...
+  inlineAttr = DictionaryAttr::getWithSorted(context, {inlineId});
+  auto flattenId = NamedAttribute(
+      Identifier::get("class", context),
+      StringAttr::get(context, "firrtl.transforms.FlattenAnnotation"));
+  flattenAttr = DictionaryAttr::getWithSorted(context, {flattenId});
+}
+
+void Inliner::run() {
+  // If the top module is not a regular module, there is nothing to do.
+  auto topModule = dyn_cast<FModuleOp>(symbolTable.lookup(circuit.name()));
```
Comment (lattner) [inline lib/Dialect/FIRRTL/Transforms/ModuleInliner.cpp:354]: CircuitOp has a helper for this, plz use it.

File: lib/Dialect/FIRRTL/Transforms/ModuleInliner.cpp [hunk 10]
```diff
@@ -0,0 +1,394 @@
...
+  if (topModule) {
+    worklist.push_back(topModule);
+  }
+  liveModules.insert(topModule);
```
Comment (lattner) [inline lib/Dialect/FIRRTL/Transforms/ModuleInliner.cpp:358]: Is this correct in the null case?

File: lib/Dialect/FIRRTL/Transforms/ModuleInliner.cpp [hunk 11]
```diff
@@ -0,0 +1,394 @@
...
+    if (shouldFlatten(module)) {
+      flattenInstances(module);
+      continue;
+    }
+    inlineInstances(module);
+  }
+
+  // Delete all unreferenced modules.
+  for (auto &op : llvm::make_early_inc_range(*circuit.getBody())) {
+    if (isa<FModuleOp, FExtModuleOp>(op) && !liveModules.count(&op)) {
```
Comment (lattner) [inline lib/Dialect/FIRRTL/Transforms/ModuleInliner.cpp:373]: Circuit can only contain modules now that the terminator is gone, so you can delete the isa check (or change it into an assert)

**Review by lattner — APPROVED** [review]
Looking great!

File: lib/Dialect/FIRRTL/Transforms/ModuleInliner.cpp (hunk 1 above)
Comment (youngar) [reply lib/Dialect/FIRRTL/Transforms/ModuleInliner.cpp]: I had a bug with this when I was working on it, took me a while to find it even though I knew what the issue was going to be.  I have updated it to use StringRefs :)

File: lib/Dialect/FIRRTL/Transforms/ModuleInliner.cpp (hunk 4 above)
Comment (youngar) [reply lib/Dialect/FIRRTL/Transforms/ModuleInliner.cpp]: Fixed!

File: lib/Dialect/FIRRTL/Transforms/ModuleInliner.cpp (hunk 5 above)
Comment (youngar) [reply lib/Dialect/FIRRTL/Transforms/ModuleInliner.cpp]: Fixed x4

File: lib/Dialect/FIRRTL/Transforms/ModuleInliner.cpp (hunk 9 above)
Comment (youngar) [reply lib/Dialect/FIRRTL/Transforms/ModuleInliner.cpp]: Fixed!  I was worried that the lookup function would construct its own symbol table, but looks like it just does a linear scan through IR.

File: lib/Dialect/FIRRTL/Transforms/ModuleInliner.cpp (hunk 10 above)
Comment (youngar) [reply lib/Dialect/FIRRTL/Transforms/ModuleInliner.cpp:345]: This was busted, I got confused at some point and thought you couldn't have an ExtModule as the main module, and I actually meant to delete this check.

Apparently that isn't remotely correct and I have fixed this check and added a test case where the top module is an extmodule.

File: lib/Dialect/FIRRTL/Transforms/ModuleInliner.cpp (hunk 2 above)
Comment (youngar) [reply lib/Dialect/FIRRTL/Transforms/ModuleInliner.cpp]: This has been mostly fixed up in issue https://github.com/llvm/circt/issues/856.  There is one more change I want to make, and have opened a PR here: https://github.com/llvm/circt/pull/964.  (this PR is based on top of that change)

**Comment by lattner** [comment]
Hey @youngar what's your plan with this patch?  Should I re-review it or something?  Why not land it?

**Comment by youngar** [comment]
@lattner I should have probably landed it a while ago.  I was just waiting until all the unresolved issues were solved and the last one remaining was flip canonicalization. I think I should do one more pass over how the wires are created for input and output ports and then i'll land it.

**Comment by lattner** [comment]
Aha got it.  If you'd like to wait for @seldridge 's output patch to land that is also fine.  I didn't realize there was a dependency, thanks!

**Comment by youngar** [comment]
I added a test which inlined a module with bundle parameters, and added some code to strip the outer flip from output ports. I think this is a good time to merge this PR since the output port semantics are settled, we are just changing the representation.

**Pull request merged**